File: Linguagem de Programação/Laboratório Semana 3/atv1/processador_texto.js
/**
 * PROCESSADOR DE TEXTO - ANÁLISE DE FREQUÊNCIA DE PALAVRAS
 * =========================================================
 * 
 * Este programa lê um arquivo de texto, processa seu conteúdo e identifica
 * as 5 palavras mais frequentes, salvando os resultados em um arquivo de saída.
 * 
 * INSTRUÇÕES CHAVE:
 * -----------------
 * 1. Leitura de arquivo com fs (File System)
 * 2. Normalização de texto (minúsculas, remoção de pontuação)
 * 3. Uso de Map para contagem de frequências
 * 4. Ordenação de array de objetos
 * 5. Escrita de resultados em arquivo de saída
 */

const fs = require('fs');

/**
 * Classe responsável por processar texto e analisar frequência de palavras.
 * 
 * ESTRUTURAS USADAS:
 * ------------------
 * - fs.readFileSync/writeFileSync: Leitura/escrita síncrona de arquivos
 * - RegExp: Expressões regulares para limpeza de texto
 * - Map: Estrutura de dados para contagem eficiente
 * - Array methods: filter, map, sort para processamento
 */
class ProcessadorTexto {
    /**
     * Inicializa o processador com o caminho do arquivo de entrada.
     * 
     * @param {string} arquivoEntrada - Caminho do arquivo a ser processado
     * 
     * INSTRUÇÃO CHAVE:
     * ----------------
     * - constructor: Método especial para inicializar objetos da classe
     * - this: Referência ao próprio objeto instanciado
     */
    constructor(arquivoEntrada) {
        this.arquivoEntrada = arquivoEntrada;
        this.texto = '';
        this.palavras = [];
        this.frequencias = new Map();
    }

    /**
     * Lê o conteúdo do arquivo de texto.
     * 
     * INSTRUÇÕES CHAVE:
     * -----------------
     * - fs.readFileSync(): Leitura síncrona (bloqueia execução até terminar)
     * - 'utf-8': Encoding para suportar caracteres especiais
     * - try-catch: Tratamento de exceções/erros
     * - return boolean: Indica sucesso ou falha da operação
     */
    lerArquivo() {
        try {
            this.texto = fs.readFileSync(this.arquivoEntrada, 'utf-8');
            console.log(`✓ Arquivo '${this.arquivoEntrada}' lido com sucesso!`);
            return true;
        } catch (erro) {
            if (erro.code === 'ENOENT') {
                console.log(`✗ Erro: Arquivo '${this.arquivoEntrada}' não encontrado!`);
            } else {
                console.log(`✗ Erro ao ler arquivo: ${erro.message}`);
            }
            return false;
        }
    }

    /**
     * Processa o texto extraindo e normalizando as palavras.
     * 
     * INSTRUÇÕES CHAVE:
     * -----------------
     * - toLowerCase(): Converte para minúsculas (normalização)
     * - match(): Usa regex para extrair palavras
     * - /\b\w+\b/g: Padrão regex para capturar palavras
     *   * \b: Limite de palavra
     *   * \w+: Um ou mais caracteres alfanuméricos
     *   * g: Flag global (todas as ocorrências)
     * - || []: Operador OR para retornar array vazio se null
     */
    processarTexto() {
        // Converte para minúsculas para contagem case-insensitive
        const textoMinusculo = this.texto.toLowerCase();
        
        // Extrai apenas palavras (remove pontuação e espaços)
        // match() retorna null se não encontrar, por isso usamos || []
        this.palavras = textoMinusculo.match(/\b\w+\b/g) || [];
        
        console.log(`✓ ${this.palavras.length} palavras extraídas do texto.`);
    }

    /**
     * Conta a frequência de cada palavra no texto.
     * 
     * ESTRUTURAS USADAS:
     * ------------------
     * - Map: Estrutura chave-valor otimizada
     *   * get(key): Obtém valor associado à chave
     *   * set(key, value): Define valor para uma chave
     * - forEach(): Itera sobre array executando função para cada elemento
     * - Arrow function: Sintaxe compacta (palavra) => { ... }
     */
    contarFrequencias() {
        // Itera sobre cada palavra e conta frequências
        this.palavras.forEach(palavra => {
            // Se palavra já existe no Map, incrementa; senão, inicia com 1
            const frequenciaAtual = this.frequencias.get(palavra) || 0;
            this.frequencias.set(palavra, frequenciaAtual + 1);
        });
        
        console.log(`✓ Frequência calculada para ${this.frequencias.size} palavras únicas.`);
    }

    /**
     * Retorna as n palavras mais frequentes.
     * 
     * @param {number} n - Número de palavras a retornar (padrão: 5)
     * @returns {Array} Array de objetos {palavra, frequencia}
     * 
     * INSTRUÇÕES CHAVE:
     * -----------------
     * - Array.from(): Converte Map em array
     * - Destructuring: [palavra, freq] extrai elementos da tupla
     * - Object literal: { palavra, freq } cria objeto (shorthand)
     * - sort(): Ordena array in-place
     * - slice(): Cria novo array com subset de elementos
     */
    obterTopPalavras(n = 5) {
        // Converte Map em array de objetos e ordena por frequência
        const palavrasOrdenadas = Array.from(this.frequencias)
            .map(([palavra, freq]) => ({ palavra, frequencia: freq }))
            .sort((a, b) => b.frequencia - a.frequencia);
        
        // Retorna apenas os n primeiros elementos
        return palavrasOrdenadas.slice(0, n);
    }

    /**
     * Gera um relatório com os resultados e salva em arquivo.
     * 
     * @param {string} arquivoSaida - Nome do arquivo de saída
     * 
     * INSTRUÇÕES CHAVE:
     * -----------------
     * - Template literals: `texto ${variavel}` para interpolação
     * - Array.join('\n'): Junta elementos com quebra de linha
     * - padEnd/padStart: Métodos para padding e alinhamento
     * - fs.writeFileSync(): Escrita síncrona em arquivo
     */
    gerarRelatorio(arquivoSaida = 'resultado_frequencias.txt') {
        const top5 = this.obterTopPalavras(5);
        
        // Prepara o conteúdo do relatório
        const relatorio = [];
        relatorio.push('='.repeat(60));
        relatorio.push('ANÁLISE DE FREQUÊNCIA DE PALAVRAS');
        relatorio.push('='.repeat(60));
        relatorio.push(`\nArquivo analisado: ${this.arquivoEntrada}`);
        relatorio.push(`Total de palavras: ${this.palavras.length}`);
        relatorio.push(`Palavras únicas: ${this.frequencias.size}`);
        relatorio.push('\n' + '-'.repeat(60));
        relatorio.push('TOP 5 PALAVRAS MAIS FREQUENTES');
        relatorio.push('-'.repeat(60));
        relatorio.push(
            'Posição'.padEnd(10) + 
            'Palavra'.padEnd(20) + 
            'Frequência'.padStart(10)
        );
        relatorio.push('-'.repeat(60));
        
        // Adiciona cada palavra do top 5
        top5.forEach((item, index) => {
            const posicao = String(index + 1).padEnd(10);
            const palavra = item.palavra.padEnd(20);
            const freq = String(item.frequencia).padStart(10);
            relatorio.push(posicao + palavra + freq);
        });
        
        relatorio.push('='.repeat(60));
        
        // Junta todas as linhas com quebra de linha
        const conteudoRelatorio = relatorio.join('\n');
        
        // Salva em arquivo
        try {
            fs.writeFileSync(arquivoSaida, conteudoRelatorio, 'utf-8');
            console.log(`\n✓ Relatório salvo em '${arquivoSaida}'`);
        } catch (erro) {
            console.log(`\n✗ Erro ao salvar relatório: ${erro.message}`);
        }
        
        // Também imprime no console
        console.log('\n' + conteudoRelatorio);
        
        return conteudoRelatorio;
    }
}

/**
 * Função principal que coordena a execução do programa.
 * 
 * FLUXO DO PROGRAMA:
 * ------------------
 * 1. Criar instância do processador
 * 2. Ler arquivo de entrada
 * 3. Processar e normalizar texto
 * 4. Contar frequências
 * 5. Gerar e exibir relatório
 */
function main() {
    console.log('\n' + '='.repeat(60));
    console.log('PROCESSADOR DE TEXTO - ANÁLISE DE FREQUÊNCIA');
    console.log('='.repeat(60) + '\n');
    
    // Caminho do arquivo de entrada
    const arquivoEntrada = 'texto.txt';
    
    // Cria instância do processador
    const processador = new ProcessadorTexto(arquivoEntrada);
    
    // Executa o processamento
    if (processador.lerArquivo()) {
        processador.processarTexto();
        processador.contarFrequencias();
        processador.gerarRelatorio();
    } else {
        console.log('\n✗ Falha no processamento. Verifique o arquivo de entrada.');
    }
}

/**
 * PONTO DE ENTRADA DO PROGRAMA
 * 
 * INSTRUÇÃO CHAVE:
 * ----------------
 * - Node.js executa o código do arquivo automaticamente
 * - Verifica se está sendo executado diretamente (não importado como módulo)
 * - require.main === module: Similar ao if __name__ == "__main__" do Python
 */
if (require.main === module) {
    main();
}

// Exporta a classe para uso como módulo (opcional)
module.exports = ProcessadorTexto;

File: Linguagem de Programação/Laboratório Semana 3/atv1/processador_texto.py
"""
PROCESSADOR DE TEXTO - ANÁLISE DE FREQUÊNCIA DE PALAVRAS
=========================================================

Este programa lê um arquivo de texto, processa seu conteúdo e identifica
as 5 palavras mais frequentes, salvando os resultados em um arquivo de saída.

INSTRUÇÕES CHAVE:
-----------------
1. Leitura de arquivo com encoding UTF-8
2. Normalização de texto (minúsculas, remoção de pontuação)
3. Uso de dicionário para contagem de frequências
4. Ordenação de dicionário por valores
5. Escrita de resultados em arquivo de saída
"""

import re
from collections import Counter
import os


class ProcessadorTexto:
    """
    Classe responsável por processar texto e analisar frequência de palavras.
    
    ESTRUTURAS USADAS:
    ------------------
    - re: Módulo de expressões regulares para limpeza de texto
    - Counter: Estrutura de dados especializada em contagem
    - Dictionary comprehension: Para processamento eficiente
    """
    
    def __init__(self, arquivo_entrada):
        """
        Inicializa o processador com o caminho do arquivo de entrada.
        
        Args:
            arquivo_entrada (str): Caminho do arquivo a ser processado
        """
        self.arquivo_entrada = arquivo_entrada
        self.texto = ""
        self.palavras = []
        self.frequencias = {}
    
    def ler_arquivo(self):
        """
        Lê o conteúdo do arquivo de texto.
        
        INSTRUÇÕES CHAVE:
        -----------------
        - with: Gerenciador de contexto que fecha o arquivo automaticamente
        - encoding='utf-8': Suporte a caracteres especiais (acentos, etc)
        - read(): Lê todo o conteúdo do arquivo de uma vez
        """
        try:
            with open(self.arquivo_entrada, 'r', encoding='utf-8') as arquivo:
                self.texto = arquivo.read()
            print(f"✓ Arquivo '{self.arquivo_entrada}' lido com sucesso!")
            return True
        except FileNotFoundError:
            print(f"✗ Erro: Arquivo '{self.arquivo_entrada}' não encontrado!")
            return False
        except Exception as e:
            print(f"✗ Erro ao ler arquivo: {e}")
            return False
    
    def processar_texto(self):
        """
        Processa o texto extraindo e normalizando as palavras.
        
        INSTRUÇÕES CHAVE:
        -----------------
        - lower(): Converte todo texto para minúsculas (normalização)
        - re.findall(): Usa regex para extrair apenas palavras (remove pontuação)
        - r'\b\w+\b': Padrão regex que captura palavras completas
          * \b: Limite de palavra
          * \w+: Um ou mais caracteres alfanuméricos
        """
        # Converte para minúsculas para contagem case-insensitive
        texto_minusculo = self.texto.lower()
        
        # Extrai apenas palavras (remove pontuação e espaços)
        self.palavras = re.findall(r'\b\w+\b', texto_minusculo)
        
        print(f"✓ {len(self.palavras)} palavras extraídas do texto.")
    
    def contar_frequencias(self):
        """
        Conta a frequência de cada palavra no texto.
        
        ESTRUTURAS USADAS:
        ------------------
        - Counter: Classe especializada da biblioteca collections
          * Subclasse de dict otimizada para contagem
          * most_common(n): Retorna as n palavras mais frequentes
        - dict(): Conversão explícita para dicionário padrão
        """
        # Counter é mais eficiente que um loop manual para contagem
        contador = Counter(self.palavras)
        self.frequencias = dict(contador)
        
        print(f"✓ Frequência calculada para {len(self.frequencias)} palavras únicas.")
    
    def obter_top_palavras(self, n=5):
        """
        Retorna as n palavras mais frequentes.
        
        Args:
            n (int): Número de palavras a retornar (padrão: 5)
        
        Returns:
            list: Lista de tuplas (palavra, frequência)
        
        INSTRUÇÕES CHAVE:
        -----------------
        - sorted(): Função built-in para ordenação
        - key=lambda: Função anônima para especificar critério de ordenação
        - x[1]: Ordena pelo segundo elemento da tupla (frequência)
        - reverse=True: Ordem decrescente (maior para menor)
        - [:n]: Slice para pegar apenas os n primeiros
        """
        # Ordena o dicionário por valores (frequências) em ordem decrescente
        palavras_ordenadas = sorted(
            self.frequencias.items(),
            key=lambda x: x[1],
            reverse=True
        )
        
        return palavras_ordenadas[:n]
    
    def gerar_relatorio(self, arquivo_saida='resultado_frequencias.txt'):
        """
        Gera um relatório com os resultados e salva em arquivo.
        
        Args:
            arquivo_saida (str): Nome do arquivo de saída
        
        INSTRUÇÕES CHAVE:
        -----------------
        - f-strings: Formatação de strings moderna (f"{variavel}")
        - write(): Escreve string no arquivo
        - \n: Caractere de nova linha
        - Formatting com :>3d e :<20s para alinhamento
        """
        top_5 = self.obter_top_palavras(5)
        
        # Prepara o conteúdo do relatório
        relatorio = []
        relatorio.append("=" * 60)
        relatorio.append("ANÁLISE DE FREQUÊNCIA DE PALAVRAS")
        relatorio.append("=" * 60)
        relatorio.append(f"\nArquivo analisado: {self.arquivo_entrada}")
        relatorio.append(f"Total de palavras: {len(self.palavras)}")
        relatorio.append(f"Palavras únicas: {len(self.frequencias)}")
        relatorio.append("\n" + "-" * 60)
        relatorio.append("TOP 5 PALAVRAS MAIS FREQUENTES")
        relatorio.append("-" * 60)
        relatorio.append(f"{'Posição':<10} {'Palavra':<20} {'Frequência':>10}")
        relatorio.append("-" * 60)
        
        for i, (palavra, freq) in enumerate(top_5, 1):
            relatorio.append(f"{i:<10} {palavra:<20} {freq:>10}")
        
        relatorio.append("=" * 60)
        
        # Junta todas as linhas com quebra de linha
        conteudo_relatorio = "\n".join(relatorio)
        
        # Salva em arquivo
        try:
            with open(arquivo_saida, 'w', encoding='utf-8') as arquivo:
                arquivo.write(conteudo_relatorio)
            print(f"\n✓ Relatório salvo em '{arquivo_saida}'")
        except Exception as e:
            print(f"\n✗ Erro ao salvar relatório: {e}")
        
        # Também imprime no console
        print("\n" + conteudo_relatorio)
        
        return conteudo_relatorio


def main():
    """
    Função principal que coordena a execução do programa.
    
    FLUXO DO PROGRAMA:
    ------------------
    1. Criar instância do processador
    2. Ler arquivo de entrada
    3. Processar e normalizar texto
    4. Contar frequências
    5. Gerar e exibir relatório
    """
    print("\n" + "=" * 60)
    print("PROCESSADOR DE TEXTO - ANÁLISE DE FREQUÊNCIA")
    print("=" * 60 + "\n")
    
    # Caminho do arquivo de entrada
    arquivo_entrada = 'texto.txt'
    
    # Cria instância do processador
    processador = ProcessadorTexto(arquivo_entrada)
    
    # Executa o processamento
    if processador.ler_arquivo():
        processador.processar_texto()
        processador.contar_frequencias()
        processador.gerar_relatorio()
    else:
        print("\n✗ Falha no processamento. Verifique o arquivo de entrada.")


if __name__ == "__main__":
    """
    Ponto de entrada do programa.
    
    INSTRUÇÃO CHAVE:
    ----------------
    - if __name__ == "__main__": Garante que main() só execute quando
      o script é executado diretamente (não quando importado como módulo)
    """
    main()
